### ISSUE TITLE
Implement break categories

### ISSUE THREAD

**Issue by czlee** [czlee/tabbycat#119]
Currently Open and ESL breaks are hard-coded. Ideally divisions should be fully configurable.

**Comment by czlee** [comment]
Requirements
----------------------
* Should be able to navigate mutual exclusion between break categories and institution caps
* Should support highlighting or labeling in relevant places

Fields
-----------------
* `tournament` - `ForeignKey` to `Tournament`
* `name` - human-readable name, e.g. "Open"
* `slug` - code-compatible name, e.g. "open"
* `seq` - the order in which they will be displayed in the user interface
* `break_size` - number of teams in this break
* `highlight_color` - colour to use when highlighting teams in the adj allocation interface(?)
* `institution_cap` - maximum number of teams that can break in this category from a single institution
* `priority` - a priority order for mutually exclusive categories, categories that can be broken into all at the same time should have the same priority, lower numbers are higher priorities
    * example 1: There are open, ESL and EFL categories, and EFL-eligible teams can break into all of them. They should all have priority 1.
    * example 2: There are open, ESL and EFL categories, but ESL-eligible teams breaking into the open category may only compete in the open break, and EFL-eligible teams breaking into the ESL break may only compete in the ESL break. Open should have priority 1, ESL 2 and EFL 3.
    * example 3: There are open, ESL and novice categories. Teams breaking into the open break may only compete in the open break, but teams not in the open break may compete in ESL and novice simultaneously. Open should have priority 1, ESL 2 and novice 2.

Teams should have a `ManyToManyField` pointing to `BreakCategory`.

Where used
----------------------
* The breaking teams page should be made generic.
* The break menu should be autogenerated based on the tournament's break categories.
* Eligibility should be indicated on adj allocation and participant pages.

Remarks
-----------------
* This would supersede the `cannot_break` field. Teams that cannot break, *e.g.* swing teams, should just not be in any break category.
* Ideally, this would supersede the `esl` and `efl` fields, though it depending on how they're used, it might be worth writing properties of `Team` to return a `bool` based on the break categories.
* I'd say that this might supersede `type` as well, but I'm not really sure how this is currently used.

@philipbelesky, any thoughts? Especially re `highlight_color`, is it worth making this generically specifiable?

**Comment by philipbelesky** [comment]
Sounds good, although I'm not sure how generalisable it needs to be. I can't think of any other break categories that would come up apart from ESL and EFL?

A side note that interacts with this: preferably break teams (for all categories) should have to be manually 'checked in' before that round is generated. For example at smaller tournaments a team might not be available for the break rounds (or subsequent break rounds). And for WADL's purposes the breaking teams are determined by top teams per division + a few other rules, so there's a preference for allowing a degree of manual control. 

RE ```highlight_color``` is this only being used in the adj allocation interface? Highlights for language status are being worked on at the moment, should be done later today

**Comment by czlee** [comment]
Fair question. Novice breaks are common in America, so there's at least one. But the more I thought about it, the more I realized it's not predicated on how many break categories exist. Once there are more than about two (and maybe even if there are two), it makes sense to abstract out the code managing them just as a DRY thing. Then our options are:
(a) Have a bunch of preset breaks that you can enable or disable in config.
(b) Have no presets, and just have a variable-length list somewhere.

I suspect (b) will be cleaner to code, since (a) requires a bunch of if statements to filter for each preset break.

I'm a bit confused about the side note. As in, the break should not be generated until the tab director is forced to verify the break-eligible teams list one last time? Agree re manual control. Well I mean, either we could try and work these rules into the flexibility, or would the ability to edit match-ups in the draft draw be sufficient? Or would we need an interface to manually check/uncheck breaking teams, similar to breaking adjudicators?

Re `highlight_color`, I just mean, is it a useful thing to put in the database (*i.e.* configuration)? Or would it overcomplicate the highlighting stuff, in such a way that we should find some other solution?

On another note, in the long run, I guess a UI to configure all this should be somewhere in the configuration pages, after they're revamped as per #139.

**Comment by czlee** [comment]
Done. @philipbelesky, if you want to review the template code that'd probably be a good idea.

Also, there's a "break eligibility" page under the Break menu in admin, for the purpose of setting break eligibility.

Known limitations:
1. I implemented an `is_general` switch. This is meant to be set to `True` for categories that most teams are eligible for, i.e. the open break, so that the UI doesn't unnecessarily clutter itself with "(Open)" everywhere. But it doesn't work in reverse: it will not show teams that cannot break in the open break as "(Not Open)" or anything like that. I'm not sure how important this is, but it'd be nice.
2. Relatedly, I removed the `cannot_break` switch, as this is in principle implemented by deselecting all break categories for that team. However, this means "ineligible" teams will not show in the breaking teams page, as the page doesn't even know they exist (the filter happens too early). Teams capped out by the institution cap still show as such.
3. I haven't tested the importer, but it probably won't be needed, there are few enough categories that they can be added by hand.
